Calvinjk/USS-Nebula
# UN-27 Procedural generator

This actually will cover two tickets:
https://uss-nebula.atlassian.net/browse/UN-27
https://uss-nebula.atlassian.net/browse/UN-58

Things I changed/added:

Created a new folder - Assets/Scenes/Testing Scenes
2 new prefabs:
- FloorTile
- GameManager

4 new scripts: 
- Tile.cs
- Map.cs
- GameManager.cs
- DungeonMapGenerator.cs

How to test:
Once you open Unity, head to Assets/Scenes/Testing Scenes and open up the scene `generator-test`.  Since I don't have any nice UI elements set up for this, you can use this scene as follows:
Once you click on play, clicking on the `Generate New Dungeon` button in the Game tab will create a new 50x50 dungeon and wipe any old generated assets.  There isn't a nice camera set up, so in order to see what it generated you will have to explore using the Scene tab.  I recommend splitting these tabs like below so you can see:
![image](https://user-images.githubusercontent.com/8600069/145096954-4002fd62-237e-4360-bde4-a9dd2337abc8.png)

Currently I am just using colored rectangles to denote what is what, we can fill those in with sprites/models later.  
- Yellow = Ungenerated tiles that are within the given boundaries
- Magenta = Doors
- White = Room Floors
- Grey = Walls

Verify that every room is reachable and nothing looks "weird".   I also had a bug for awhile that re-generating a map caused unity to crash, but I believe it is no longer an issue.  Let me know if this happens while you are messing around.

To use the generator in a new scene, all you have to do is drop the GameManager into that scene.  Then call GameManager.GenerateNewMap(int size) where the generated map will be size x size in dimensions.  In the example scene I did this through a button's OnClick() function.

KNOWN ISSUES:
- The algorithm is inefficient.  I wrote this thing a long time ago and its time complexity is awful.  I will make a ticket for making this thing more efficient, but we are going to ignore that for now.  As a consequence of this, generating anything above 100x100 will be slow, and I wouldn't recommend going higher than 300x300 at all.
- This currently creates some unnecessary artifacts that we likely don't need, but might be helpful for now (ungenerated tiles).  We will remove these later, but for now its a good visualization of how the algorithm is working.
- There are a lot of custom variables in this script (min/max room size, max attempts, etc) that are not currently able to be modified by calling the top-level GenerateNewMap().   Eventually we will want to expose these, but I'll make another ticket for that.
